File: src/boxing_coach/analysis/context.py
"""AnalysisContext — the single object every rule receives.

It carries the raw pose sequence, the drill context, and the shared derived
features (body scale, punch events) so each rule pulls exactly what it needs
and expensive work (punch detection) happens once, not per rule.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from functools import cached_property

from ..domain.drill import DrillContext
from ..domain.landmarks import Side
from ..domain.pose import PoseSequence
import numpy as np

from .features import PunchDetector, PunchDetectorConfig, PunchEvent, compute_body_scale
from .round_profile import RoundProfile, compute_round_profile, stance_speed_series
from .style import DEFAULT_STYLE_PROFILE, StyleProfile


@dataclass
class AnalysisContext:
    """Everything a rule needs to evaluate one round."""

    sequence: PoseSequence
    drill: DrillContext
    punch_config: PunchDetectorConfig = field(default_factory=PunchDetectorConfig)
    #: Tunes/gates the rules for this round's fighting style. Defaults to the
    #: neutral high-guard profile so tests and callers that don't care are
    #: unaffected. The adapter resolves it from `drill.style`.
    style_profile: StyleProfile = DEFAULT_STYLE_PROFILE

    @cached_property
    def body_scale(self) -> float:
        return compute_body_scale(self.sequence)

    @cached_property
    def punches(self) -> list[PunchEvent]:
        return PunchDetector(self.punch_config).detect(self.sequence, self.body_scale)

    @cached_property
    def round_profile(self) -> RoundProfile:
        return compute_round_profile(self.sequence, self.punches, self.body_scale)

    @cached_property
    def stance_speed(self) -> np.ndarray:
        """Per-frame stance-centre speed (torso-lengths/sec); the in/out signal.

        Index-aligned to `sequence.frames`. Computed once here so the guard
        rules that read it (excusing a low hand while the feet step in/out)
        share the work.
        """
        return stance_speed_series(self.sequence, self.body_scale)

    def punches_by(self, side: Side) -> list[PunchEvent]:
        return [p for p in self.punches if p.side is side]

File: app/lib/analysis/context.dart
import 'drill.dart';
import 'features.dart';
import 'landmarks.dart';
import 'pose.dart';
import 'style.dart';

/// AnalysisContext — the single object every rule receives. Mirror of
/// `src/boxing_coach/analysis/context.py`.
///
/// It carries the raw pose sequence, the drill context, and the shared derived
/// features (body scale, punches, stance speed) so each rule pulls exactly what
/// it needs and the expensive work happens once, not per rule. The derived
/// features are computed lazily and cached, matching Python's `cached_property`.
class AnalysisContext {
  AnalysisContext({
    required this.sequence,
    required this.drill,
    PunchDetectorConfig? punchConfig,
    this.styleProfile = defaultStyleProfile,
  }) : punchConfig = punchConfig ?? const PunchDetectorConfig();

  final PoseSequence sequence;
  final DrillContext drill;
  final PunchDetectorConfig punchConfig;

  /// Tunes/gates the rules for this round's style. Defaults to the neutral
  /// high-guard profile so callers that don't care are unaffected.
  final StyleProfile styleProfile;

  double? _bodyScale;
  List<PunchEvent>? _punches;
  List<double>? _stanceSpeed;

  double get bodyScale => _bodyScale ??= computeBodyScale(sequence);

  List<PunchEvent> get punches =>
      _punches ??= PunchDetector(config: punchConfig).detect(sequence, bodyScale);

  /// Per-frame stance-centre speed (torso-lengths/sec); the in/out signal,
  /// index-aligned to `sequence.frames`.
  List<double> get stanceSpeed =>
      _stanceSpeed ??= stanceSpeedSeries(sequence, bodyScale);

  List<PunchEvent> punchesBy(Side side) =>
      punches.where((p) => p.side == side).toList();
}
